src/lib/operators/join_mpsm.cpp

```diff
@@ -0,0 +1,520 @@
+                  const PredicateCondition op, JoinMode mode)
+      : _mpsm_join{sort_merge_join},
+        _left_column_id{left_column_id},
+        _right_column_id{right_column_id},
+        _op{op},
+        _mode{mode} {
+    _cluster_count = _determine_number_of_clusters();
+    _output_pos_lists_left.resize(_cluster_count);
+    _output_pos_lists_right.resize(_cluster_count);
+    for (size_t cluster = 0; cluster < _cluster_count; ++cluster) {
+      _output_pos_lists_left[cluster].resize(1);
+      _output_pos_lists_right[cluster].resize(_cluster_count);
+    }
+  }
+
+ protected:
+  JoinMPSM& _mpsm_join;
+
+  // Contains the materialized sorted input tables
+  std::unique_ptr<MaterializedNUMAPartitionList<T>> _sorted_left_table;
+  std::unique_ptr<MaterializedNUMAPartitionList<T>> _sorted_right_table;
+
+  // Contains the null value row ids if a join column is an outer join column
+  std::unique_ptr<PosList> _null_rows_left;
+  std::unique_ptr<PosList> _null_rows_right;
+
+  const ColumnID _left_column_id;
+  const ColumnID _right_column_id;
+
+  const PredicateCondition _op;
+  const JoinMode _mode;
+
+  // the cluster count must be a power of two, i.e. 1, 2, 4, 8, 16, ...
+  ClusterID _cluster_count;
+
+  // Contains the output row ids for each cluster
+  std::vector<std::vector<std::shared_ptr<PosList>>> _output_pos_lists_left;
+  std::vector<std::vector<std::shared_ptr<PosList>>> _output_pos_lists_right;
+
+  /**
+   * The TablePosition is a utility struct that is used to define a specific position in a sorted input table.
+  **/
+  struct TableRange;
+  struct TablePosition {
+    TablePosition() {}
+    TablePosition(NodeID partition, ClusterID cluster, size_t index)
+        : cluster{cluster}, index{index}, partition{partition} {}
+
+    ClusterID cluster;
+    size_t index;
+    NodeID partition;
+
+    TableRange to(TablePosition position) { return TableRange(*this, position); }
+  };
+
+  /**
+    * The TableRange is a utility struct that is used to define ranges of rows in a sorted input table spanning from
+    * a start position to an end position.
+  **/
+  struct TableRange {
+    TableRange(TablePosition start_position, TablePosition end_position) : start{start_position}, end{end_position} {
+      DebugAssert(start.partition == end.partition, "Table ranges are only allowed over the same position");
+    }
+    TableRange(NodeID partition, ClusterID cluster, size_t start_index, size_t end_index)
+        : start{TablePosition(partition, cluster, start_index)}, end{TablePosition(partition, cluster, end_index)} {
+      DebugAssert(start.partition == end.partition, "Table ranges are only allowed over the same position");
+    }
+
+    TablePosition start;
+    TablePosition end;
+
+    // Executes the given action for every row id of the table in this range.
+    template <typename F>
+    void for_every_row_id(std::unique_ptr<MaterializedNUMAPartitionList<T>>& table, F action) {
```

Comment:
A high-level comment on how you implement NUMA awareness would be great. How do you decide how to partition things, how and where do you make sure things are on the correct NUMA node and what is the NUMA layout of the result?